### ISSUE TITLE
[TODO] make openArray first-class

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/Nim#9001]
I didn't find an issue to track this, it's only mentioned in comments in random places, eg [1][2].

## proposal

allow:
```nim
let (pr, n) : (ptr int, int) = get_pointer_length(...) # eg, from C proc, etc
let a=toOpenArray(pr, 0, n-1) # a is openArray[int]
```

It could be enabled via a `-d:unsafe_openArray_firstclass` (and assert at runtime with this flag is not passed)

## links

[1] https://github.com/nim-lang/Nim/pull/5957#issuecomment-356895299

> Stylistic issues aside, it's mostly fine but I think it would be ill advised to introduce this. Once openArray and var openArray work with some minor Rust inspired borrow checking to make it safe (RFC soon to be written) there would be considerable overlap with this stdlib module.

[2] https://github.com/nitely/nim-graphemes/issues/4
> I'll have to wait until OpenArray can be used as a return value

* ICE's and bugs with openarray #8259

**Comment by timotheecour** [comment]
hence the `-d:unsafe_openArray_firstclass` which means: the programmer knows what he's doing, he's responsible for making sure memory isn't corrupted and using it at his own risk, just like `cast[T]`; it enables valid use cases in particular for performance reasons, bypassing copies.

**Comment by mratsim** [comment]
Duplicate of closed https://github.com/nim-lang/Nim/issues/5437

Discussions with several proposals to read: https://github.com/nim-lang/Nim/issues/7337#issuecomment-373444925. (in the context of `openarray[byte]`/raw memory blob)

I feel like we are closing Openarrays/Views/Slices/Ranges/Spans feature requests very early.

**Comment by timotheecour** [comment]
> If he knows what he is doing he is free to use ptr UncheckedArray[T] already.

`UncheckedArray[T]` doesn't have a `len` field, so it's not a suitable replacement to all the proposals that have been made so far regarding having a `Openarrays/Views/Slices/Ranges/Spans`

**Comment by Araq** [comment]
> UncheckedArray[T] doesn't have a len field, so it's not a suitable replacement to all the proposals that have been made so far regarding having a Openarrays/Views/Slices/Ranges/Spans

Well but you want to slice into the data so you want control over the length too.

**Comment by timotheecour** [comment]
> Well but you want to slice into the data so you want control over the length too.

yes, but as in all (or most?) proposals that I've seen, the length should be part of the object returned by toOpenArray / toSpan / etc:

```
let (pr, n) : (ptr int, int) = get_pointer_length(...) # eg, from C proc, etc
# with 1st class openArray:
let a=toOpenArray(pr, 0, n-1) # a is openArray[int]
doAssert a.len == n

# alternatively, if openArray can't be made 1st class for some reason: with Span:
let a=toSpan(pr, 0, n-1) # a is Span[int]
doAssert a.len == n
```

**Comment by timotheecour** [comment]
/cc @araq 
can we re-open? it was closed automatically by github's (over-reaching IMO) heuristic due to the following wording `I don't know if that improves the ergonomics enough to close #9001, but it is a pretty simple change maybe worth consideration` in https://github.com/nim-lang/Nim/pull/9316#issue-222165110 by @c-blake
